### PR TITLE
8278833: riscv: Remove the x3 and x4 register saving logic in register savers

### PR DESCRIPTION
**Pull request by zhengxiaolinX**
Hi team,

x3 is the global pointer register used to reference global variables in C++ code, generated by linker; and x4 is the thread pointer register and would be modified by the kernel. In the riscv-port repo, we have no other modifications for the x3 & x4 registers so I would recommend removing the x3 & x4 saving logic in register savers. Previously verified along with other patches under full tiers jdk/hotspot on boards, and verified a simple `test/jtreg/hotspot/compiler` after fixing conflicts with [JDK-8278337](https://github.com/openjdk/riscv-port/pull/25).

BTW since we remove both two registers together, sp remains 16-byte aligned without other changes in stack frames.

Thanks,
Xiaolin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278833](https://bugs.openjdk.java.net/browse/JDK-8278833): riscv: Remove the x3 and x4 register saving logic in register savers

### Reviewers
 * [Yadong Wang](https://openjdk.java.net/census#yadongwang) (@yadongw - Author) ⚠️ Review applies to bdb5fba36ecdbe051b5e58df3e1a48856e138b3e
 * [Feilong Jiang](https://openjdk.java.net/census#fjiang) (@feilongjiang - Author)
 * [Fei Yang](https://openjdk.java.net/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/riscv-port pull/31/head:pull/31` \
`$ git checkout pull/31`

Update a local copy of the PR: \
`$ git checkout pull/31` \
`$ git pull https://git.openjdk.java.net/riscv-port pull/31/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 31`

View PR using the GUI difftool: \
`$ git pr show -t 31`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/riscv-port/pull/31.diff">https://git.openjdk.java.net/riscv-port/pull/31.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp [hunk 1]
```diff
@@ -287,7 +287,7 @@ static OopMap* save_live_registers(StubAssembler* sasm,
   __ block_comment("save_live_registers");
 
   // if the number of pushed regs is odd, zr will be added
```
Comment (yadongw) [inline src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: `// if the number of pushed regs is odd, zr will be added` is out-of-date after JDK-8278337.

File: src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp [hunk 2]
```diff
@@ -349,11 +349,11 @@ void Runtime1::initialize_pd() {
     sp_offset += step;
   }
 
-  // we save x0, x3 ~ x31, except x1, x2
+  // we save x0, x5 ~ x31, except x1, x2, x3, x4
```
Comment (yadongw) [inline src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: Are we still saving x0 now?

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp [hunk 3]
```diff
@@ -190,10 +190,10 @@ OopMap* RegisterSaver::save_live_registers(MacroAssembler* masm, int additional_
...
   }
 
-  // ignore zr, ra and sp, being ignored also by push_CPU_state (pushing zr only for stack alignment)
+  // ignore zr, ra, sp, gp and tp, being ignored also by push_CPU_state (pushing zr only for stack alignment)
   sp_offset_in_slots += RegisterImpl::max_slots_per_register;
   step_in_slots = RegisterImpl::max_slots_per_register;
-  for (int i = 3; i < RegisterImpl::number_of_registers; i++, sp_offset_in_slots += step_in_slots) {
+  for (int i = 5; i < RegisterImpl::number_of_registers; i++, sp_offset_in_slots += step_in_slots) {
     Register r = as_Register(i);
     if (r != xthread && r != t0 && r != t1) {
```
Comment (yadongw) [inline src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp]: Why not start the loop from x7?

File: src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp (hunk 1 above)
Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: Right -- changed.

File: src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp (hunk 2 above)
Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: No -- changed.

File: src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp (hunk 3 above)
Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp]: Thanks for the nice reminder -- changed.

Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp]: Need some testing work for the new change -- I will ping this thread after tests are done.

Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp]: `test/jtreg/hotspot/compiler` on Qemu reveals no other errors.

**Review by yadongw — APPROVED** [review]
lgtm

File: src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp [hunk 4]
```diff
@@ -332,9 +332,9 @@ static void restore_live_registers_except_r10(StubAssembler* sasm, bool restore_
     __ addi(sp, sp, 32 * wordSize);
   }
 
-  // if the number of popped regs is odd, zr will be added
-  // integer registers except ra(x1) & sp(x2) & x10
-  __ pop_reg(RegSet::range(x3, x9), sp);   // pop zr, x3 ~ x9
+  // if the number of popped regs is odd, the reserved slot for alignment will be removed
+  // integer registers except ra(x1) & sp(x2) & gp(x3) & tp(x4) & x10
+  __ pop_reg(RegSet::range(x5, x9), sp);   // pop zr, x5 ~ x9
```
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: Is `zr` will be poped?

Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: Currently it seems not: it looks like [JDK-8278337](https://github.com/openjdk/riscv-port/pull/25) only removes the zr saving and restoring logic, but just remains that stack slot originally for zr unchanged.

Comment (feilongjiang) [reply src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: comment should be updated too though.

Comment (zhengxiaolinX) [reply src/hotspot/cpu/riscv/c1_Runtime1_riscv.cpp]: Thanks for pointing it out -- seems there are unfixed comments at large out of my sight. I have fixed them according to the assembly code.

**Review by feilongjiang — APPROVED** [review]

**Comment by feilongjiang** [comment]
LGTM

**Review by RealFYang — APPROVED** [review]

**Comment by zhengxiaolinX** [comment]
Thanks for the reviews, team.

/integrate

**Pull request closed without merging**